proposals/0000-custom-executors.md

```diff
@@ -0,0 +1,464 @@
+  init<T: SerialExecutor>(serialExecutor: T, supportsSwitching: Bool)
+  init<T: Executor>(concurrentExecutor: T)
+
+  static var defaultConcurrent: UnownedExecutorRef { get }
+
+  var isSerial: Bool { get }
+
+  func asOwned() -> Executor
+}
+```
+
+### The default global concurrent executor
+
+The default concurrent executor is used to run jobs that don't
+need to run somewhere more specific. It is based on a fixed-width
+thread pool that scales to the number of available cores. Programmers
+therefore do not need to worry that creating too many jobs at once
+will cause a thread explosion that will starve the program of
+resources.
+
+Some environments may wish to fully replace the default concurrent
+executor. (For example, they may have their own fixed-width
+thread pool, and having two fixed-width pools in use at once
+completely defeats the purpose.) Programs should not attempt
+to do this by specifying a custom executor on absolutely
+everything in the program; that would be both highly invasive
+and doomed to failure. We think it should be possible to directly
+support hooking the default concurrent executor, perhaps by
+overriding a weak symbol, the same way that C programs can
+override `malloc`.  We won't discuss this further in this proposal
+because it doesn't otherwise impact the language and library design.
+
+### The default serial executor implementation
+
+The default serial executor implementation is separately instantiated
+for each actor that doesn't declare a custom executor; see "Actor
+executors" below. It is based on an "asynchronous lock" design which
+allows existing threads to immediately begin executing code on behalf
+of the actor rather than requiring functions to suspend and resume
+executing asynchronously. This process is called *switching*. In
+situations where switching is impossible, such as when the actor is
+already executing on a thread, a job to process the actor will be
+scheduled onto the default concurrent executor.
+
+Switching significantly complicates the design of serial executors,
+which can be seen in the `SerialExecutor` protocol below. This
+complexity could be eliminated by only allowing default serial
+executors to participate in switching. Currently, we are reluctant
+to hardcode this limitation into the protocol design.
+
+### Actor executors
+
+An actor's executor must conform to `SerialExecutor`, which refines
+the `Executor` protocol. `SerialExecutor` makes additional guarantees
+about the behavior of `execute(currentExecutor:)` and also adds
+several new methods which are used to allow the executor to opt in
+to supporting "switching" serial executors the same way that the
+default serial executor does.
+
+TODO: if we need to support blocking actors, we will need to extend
+this significantly; otherwise, actor executors will need to assume
+that returning from a job should always unblock the executor.
+(Presumably we do not want to block actor executors by blocking
+their thread.)
```

Comment:
do you here mean "blocking actor" as in "non-reentrant actor"?

I will read up how we execute actors today... in our other design we didn't need anything special from executors; the actor "has" a mailbox ("the lightweight queue") and submits _itself_ to an executor to run; if new tasks arrive at the actor while it is "'async' blocked" (waiting non-reentrantly on some thing), it does not submit itself to the executor, until that waiting thing comes back.

Though not sure if I understood your TODO right here, I can prototype this a bit in the coming week(s) if that'd help

Reply:
I'm just saying that we'd definitely need extra requirements in `SerialExecutor` to support locking and unlocking custom actor executors — you need the jobs that the actor is running to be able to tell it to stop processing other jobs (and then somehow restart it), because there's an abstraction layer there.

Reply:
Ok got it 👍